### PR TITLE
[c++] Remove unnecessary blob::_content member

### PR DESCRIPTION
**Pull request by ara-ayvazyan**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,8 @@ different versioning scheme, following the Haskell community's
 * Fixed `bond::DynamicParser` that may not emit transform's `OmittedField`
   for compile-time schema and an omitted optional field in the payload.
   ([Issue \#1120](https://github.com/microsoft/bond/issues/1120))
+* Removed `bond::blob`'s unnecessary `_content` member, making the type
+  smaller (32 -> 24 bytes).
```
Comment (Lyr1k) [inline CHANGELOG.md:32]: remove platform specific size?

**Review by Lyr1k — APPROVED** [review]

File: cpp/inc/bond/core/blob.h [hunk 2]
```diff
@@ -87,42 +81,36 @@ class blob
...
     /// Not recommended for performance reasons. Use boost::shared_ptr whenever possible.
     template <typename T, template <typename U> class SmartPtr>
     blob(const SmartPtr<T>& buffer, uint32_t offset, uint32_t length)
-        : _buffer(wrap_in_shared_ptr(buffer)),
-          _content(bond::detail::checked_add(_buffer.get(), offset)),
+        : _buffer(wrap_in_shared_ptr(buffer, offset)),
           _length(length)
     {
-        bond::detail::checked_add(_content, length);
+        bond::detail::checked_add(_buffer.get(), length);
```
Comment (biocomp) [inline cpp/inc/bond/core/blob.h:97]: Use `content()` or `data()` instead of `_buffer.get()` in most places?

File: cpp/inc/bond/core/blob.h [hunk 3]
```diff
@@ -416,7 +391,7 @@ inline T blob_cast(const blob& from)
 template <typename A>
 inline blob blob_prolong(blob src, const A& allocator)
 {
-    if (src._buffer)
+    if (src._buffer.use_count() != 0)
```
Comment (biocomp) [inline cpp/inc/bond/core/blob.h:394]: Is using `use_count()` safe?

**Review by biocomp — APPROVED** [review]

File: cpp/inc/bond/core/blob.h (hunk 2 above)
Comment (ara-ayvazyan) [reply cpp/inc/bond/core/blob.h:97]: Outside of the scope of this PR. The whole class needs a clean and modern rewrite.

File: cpp/inc/bond/core/blob.h (hunk 3 above)
Comment (ara-ayvazyan) [reply cpp/inc/bond/core/blob.h:394]: Here we are checking for the ownership presence (i.e. not the actual reference count). That should be safe even for concurrent access.

**Review by chwarr — CHANGES_REQUESTED** [review]
LGTM with a minor request for a comment to be added.

File: cpp/inc/bond/core/blob.h [hunk 4]
```diff
@@ -38,47 +37,42 @@ class blob
     ///
     /// Not recommended because of buffer lifetime management.
     blob(const void* content, uint32_t length)
-        : _buffer(),
-          _content(static_cast<const char*>(content)),
+        : _buffer(boost::shared_ptr<const void>(), static_cast<const char*>(content)),
```
[the remark this answers is not in the document]
Comment (ara-ayvazyan) [reply cpp/inc/bond/core/blob.h:40]: This looked obvious to me, but will add.

[the remark this answers is not in the document]
Comment (ara-ayvazyan) [reply cpp/inc/bond/core/blob.h]: Let me know if the wording of the comment looks ok.

**Pull request merged**